pkg/operator/collection.go

```diff
@@ -124,6 +125,20 @@ func newCollectionReconciler(c client.Client, opts Options) *collectionReconcile
 	}
 }
 
+func patchCollectionStatus(ctx context.Context, kubeClient client.Client, statusContainer monitoringv1.PodMonitoringStatusContainer) error {
```

Comment:
why does this have to patch the yaml manually? could it just call `Update()` with the new object?

Reply:
To reduce conflicts. Two different goroutines are now updating pod monitoring statuses at the same time:
1. The conditions.
2. The target statuses.

With an update, if it fails due to version mismatch, you have to re-create the object and try the request to update again. I believe we can avoid that with a patch.

Reply:
how does patching react to concurrent updates like this? Also if an update fails, you should be able to return an error in the [`reconcile.Result`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#Result) that re-queues the reconcile attempt and just let it run again, was this something you tried?

Reply:
In general each reconcile attempt should be idempotent, so your logic can always back off/retry and re-build its state each time. I see that none of our other control loop functions look like they're using this, so maybe there's a reason for that

Reply:
> your logic can always back off/retry and re-build its state each time
I think this is a harder issue than it seems. You need to introduce a retry amount so you don't keep retrying indefinitely and possibly with jitter, which makes updates very random. Should you re-fetch everything or should you retry with the stale fetched data? At that point, since I refresh the state every 10 seconds, the question is whether it's worth it to get the delayed update or just wait for the next one.